File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/ItemAtlasGenerator.java
package fr.hardel.asset_editor.client.javafx.lib;

import fr.hardel.asset_editor.client.rendering.ItemAtlasRenderer;
import fr.hardel.asset_editor.client.rendering.ItemAtlasRenderer.AtlasEntry;
import javafx.application.Platform;
import javafx.scene.image.PixelFormat;
import javafx.scene.image.WritableImage;
import net.minecraft.resources.Identifier;

import java.util.concurrent.CopyOnWriteArrayList;

public final class ItemAtlasGenerator {
    private static final CopyOnWriteArrayList<Runnable> listeners = new CopyOnWriteArrayList<>();
    private static volatile WritableImage atlasImage;

    static {
        ItemAtlasRenderer.subscribeGeneration(ItemAtlasGenerator::rebuild);
    }

    public static WritableImage getAtlasImage() {
        if (atlasImage == null && ItemAtlasRenderer.isReady())
            rebuild();
        return atlasImage;
    }

    public static AtlasEntry getEntry(Identifier itemId) {
        return ItemAtlasRenderer.getEntry(itemId);
    }

    public static boolean isReady() {
        return ItemAtlasRenderer.isReady();
    }

    public static Runnable subscribe(Runnable listener) {
        listeners.add(listener);
        return () -> listeners.remove(listener);
    }

    public static void rebuild() {
        int[] pixels = ItemAtlasRenderer.getArgbPixels();
        if (pixels == null)
            return;

        int w = ItemAtlasRenderer.getWidth();
        int h = ItemAtlasRenderer.getHeight();
        Runnable update = () -> {
            atlasImage = createImage(pixels, w, h);
            listeners.forEach(Runnable::run);
        };
        if (Platform.isFxApplicationThread()) update.run();
        else Platform.runLater(update);
    }

    private static WritableImage createImage(int[] argbPixels, int width, int height) {
        WritableImage image = new WritableImage(width, height);
        image.getPixelWriter().setPixels(0, 0, width, height, PixelFormat.getIntArgbInstance(), argbPixels, 0, width);
        return image;
    }

    private ItemAtlasGenerator() {}
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/ItemAtlasGenerator.kt
package fr.hardel.asset_editor.client.compose.lib

import androidx.compose.runtime.LongState
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.graphics.ImageBitmap
import fr.hardel.asset_editor.client.compose.lib.utils.argbPixelsToImageBitmap
import fr.hardel.asset_editor.client.rendering.ItemAtlasRenderer
import fr.hardel.asset_editor.client.rendering.ItemAtlasRenderer.AtlasEntry
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CopyOnWriteArrayList
import javax.swing.SwingUtilities
import net.minecraft.resources.Identifier

object ItemAtlasGenerator {

    private val listeners = CopyOnWriteArrayList<Runnable>()
    private val rebuildLock = Any()

    private val atlasState = mutableStateOf<ImageBitmap?>(null)
    private val generationState = mutableLongStateOf(-1L)

    @Volatile
    private var pendingGeneration: Long = -1L

    val atlasImageState: State<ImageBitmap?> = atlasState
    val generation: LongState = generationState

    init {
        ItemAtlasRenderer.subscribeGeneration(::rebuild)
    }

    @JvmStatic
    fun getAtlasImage(): ImageBitmap? {
        if (!ItemAtlasRenderer.isReady()) {
            ItemAtlasRenderer.requestGeneration()
            return atlasState.value
        }
        if (ItemAtlasRenderer.getGeneration() > generationState.longValue) {
            rebuild()
        }
        return atlasState.value
    }

    @JvmStatic
    fun getEntry(itemId: Identifier): AtlasEntry? {
        if (!ItemAtlasRenderer.isReady()) {
            ItemAtlasRenderer.requestGeneration()
        }
        return ItemAtlasRenderer.getEntry(itemId)
    }

    @JvmStatic
    fun subscribe(listener: Runnable): Runnable {
        listeners.add(listener)
        return Runnable { listeners.remove(listener) }
    }

    @JvmStatic
    fun rebuild() {
        val pixels = ItemAtlasRenderer.getArgbPixels() ?: return
        val width = ItemAtlasRenderer.getWidth()
        val height = ItemAtlasRenderer.getHeight()
        val generation = ItemAtlasRenderer.getGeneration()

        synchronized(rebuildLock) {
            if (generation <= generationState.longValue || generation == pendingGeneration) {
                return
            }
            pendingGeneration = generation
        }

        CompletableFuture
            .supplyAsync { argbPixelsToImageBitmap(pixels, width, height) }
            .thenAccept { image ->
                val publish = Runnable {
                    synchronized(rebuildLock) {
                        if (generation < generationState.longValue) {
                            return@Runnable
                        }
                        atlasState.value = image
                        generationState.longValue = generation
                        if (pendingGeneration == generation) {
                            pendingGeneration = -1L
                        }
                    }
                    listeners.forEach(Runnable::run)
                }

                if (SwingUtilities.isEventDispatchThread()) {
                    publish.run()
                } else {
                    SwingUtilities.invokeLater(publish)
                }
            }
            .exceptionally {
                synchronized(rebuildLock) {
                    if (pendingGeneration == generation) {
                        pendingGeneration = -1L
                    }
                }
                null
            }
    }

}
